File: tests/journal.lib/info_cascade.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>
// support
#include <cassert>


// type aliases
// chronicler
using chronicler_t = pyre::journal::chronicler_t;
// info
using myinfo_t = pyre::journal::info_t;
// the trash can
using trash_t = pyre::journal::trash_t;


// verify the cascade works correctly for infos
int main() {
    // make a channel
    myinfo_t parent("test.info.parent");
    // its activation state is what's expected
    assert(parent.state());
    // and the device is the global default
    assert(parent.device() == chronicler_t::device());
    // turn it off
    parent.deactivate();
    // and set the device to a trash can
    parent.device(std::make_shared<trash_t>());

    // make a info that's lower in the hierarchy
    myinfo_t child("test.info.parent.blah.blah.child");
    // make sure its activation state is what's expected
    assert(child.state() == parent.state());
    // and that it inherited the device
    assert(child.device() == parent.device());

    // all done
    return 0;
}


// end of file

File: tests/journal.api/info_cascade.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


def test():
    """
    Verify that channels lower in the hierarchy inherit the default state of their parent
    """
    # get the journal
    import j2

    # make a channel
    parent = j2.info(name="test.index.parent")
    # verify that the state is on
    assert parent.active is True
    # it's non-fatal
    assert parent.fatal is False
    # and the device is at the default value
    assert parent.device is j2.info.chronicler.device

    # activate it
    parent.active = True
    # make it fatal
    parent.fatal = True
    # and set the device to a j2.trash can
    parent.device = j2.trash()

    # lookup a name that is lower in the hierarchy
    child = j2.info(name="test.index.parent.blah.blah.child")
    # that it's state is the same as the parent
    assert child.active == parent.active
    assert child.fatal == parent.fatal
    # and that it inherited the device correctly
    assert child.device is parent.device

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file